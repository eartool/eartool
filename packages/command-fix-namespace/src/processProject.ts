import {
  addSingleFileReplacementsForRenames,
  processReplacements,
  ProjectContext,
  ReplacementsWrapperForContext,
  SimpleReplacements,
} from "@eartool/replacements";
import type { PackageExportRename } from "@eartool/replacements";
import { dropDtsFiles, organizeImportsOnFiles } from "@eartool/utils";
import type { PackageContext, PackageName } from "@eartool/utils";
import type { Logger } from "pino";
import { calculateNamespaceLikeRemovals } from "./calculateNamespaceLikeRemovals.js";
import { calculateNamespaceRemovals } from "./calculateNamespaceRemovals.js";

export interface Status {
  totalWorkUnits: number;
  completedWorkUnits: number;
  stage: "analyzing" | "writing" | "organizing";
}

export interface BaseOptions {
  dryRun: boolean;
  removeNamespaces: boolean;
  removeFauxNamespaces: boolean;
  organizeImports: boolean;
}

export interface ProcessProjectOpts extends BaseOptions {
  logger: Logger;
  updateState?: (data: Status) => void;
  additionalRenames?: Map<PackageName, PackageExportRename[]>;
}

/**
 * Process a project, removing namespaces and organizing imports.
 * @param project  The project to process
 * @param param1  Options
 * @returns
 */
export async function processProject(
  packageContext: PackageContext,
  {
    dryRun = false,
    logger,
    updateState = (_data) => undefined,
    additionalRenames,
    removeNamespaces,
    removeFauxNamespaces,
    organizeImports: shouldOrganizeImports,
  }: ProcessProjectOpts,
) {
  const { project } = packageContext;

  dropDtsFiles(project);
  const context = new ProjectContext(
    packageContext.project,
    logger,
    packageContext.packagePath,
    packageContext.packageName,
    packageContext.packageJson,
  );
  const totalFiles = project.getSourceFiles().length;
  // Three stages:
  // * analyzing:
  //   * if `removeNamespaces`: add `totalFiles`
  //   * if `additionalRenames`: add `totalFiles`
  // * organizing `[...changedFiles].length` work.
  // * writing  `[]
  function calculateTotalWorkUnits(changedFilesCount: number) {
    return (
      (removeNamespaces ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (dryRun ? 0 : changedFilesCount) +
      (shouldOrganizeImports ? changedFilesCount : 0)
    );
  }

  // Assume all files change for now
  let totalWorkUnits = calculateTotalWorkUnits(totalFiles);

  let completedWorkUnits = 0;
  // TODO: Rename totalFiles here to totalWorkUnits or similar
  updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });

  logger.debug("Running with opts %o", {
    dryRun,
    removeNamespaces,
    removeFauxNamespaces,
    organizeImports: shouldOrganizeImports,
    additionalRenames: [...(additionalRenames?.entries() ?? [])],
  });

  for (const sf of project.getSourceFiles()) {
    const replacements = new ReplacementsWrapperForContext(context);

    if (removeNamespaces) {
      calculateNamespaceRemovals(sf, context, replacements);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (removeFauxNamespaces) {
      calculateNamespaceLikeRemovals(sf, context, replacements);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (additionalRenames) {
      const replacements = new SimpleReplacements(logger);
      addSingleFileReplacementsForRenames(context, sf, additionalRenames, replacements, dryRun);
      for (const r of replacements.getReplacementsArray()) {
        context.addReplacement(r);
      }
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }
  }

  // actually updates files in project!
  const changedFiles = [...processReplacements(project, context.getReplacements())];
  totalWorkUnits = calculateTotalWorkUnits(changedFiles.length);

  if (shouldOrganizeImports) {
    logger.debug("Organizing imports");
    updateState({ totalWorkUnits, completedWorkUnits, stage: "organizing" });
    organizeImportsOnFiles(project, changedFiles);
  }

  completedWorkUnits += changedFiles.length; // TODO make this granular?
  updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });

  if (dryRun) {
    logger.info("DRY RUN");
  } else {
    logger.info("Saving");
    await project.save();
    completedWorkUnits += changedFiles.length; // TODO make this granular?
    updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });
  }

  logger.debug("Exported renames %o", context.getRecordedRenames());

  return {
    exportedRenames: context.getRecordedRenames(),
  };
}
