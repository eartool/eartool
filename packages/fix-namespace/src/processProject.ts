import type { Node, SourceFile } from "ts-morph";
import type { Project } from "ts-morph";
import { calculateNamespaceRemovals } from "./calculateNamespaceRemovals.js";
import type { Logger } from "pino";
import { ProjectContext } from "./Context.js";
import { processReplacements } from "./replacements/processReplacements.js";
import { dropDtsFiles } from "./utils/tsmorph/dropDtsFiles.js";
import { organizeImportsOnFiles } from "./utils/tsmorph/organizeImportsOnFiles.js";
import type { PackageExportRename } from "./replacements/PackageExportRename.js";
import type { PackageName } from "./PackageName.js";
import { addSingleFileReplacementsForRenames } from "./replacements/addSingleFileReplacementsForRenames.js";
import type { Replacement } from "./replacements/Replacement.js";
import { ReplacementsWrapper } from "./ReplacementsWrapper.js";
import { calculateNamespaceLikeRemovals } from "./calculateNamespaceLikeRemovals.js";

export interface Status {
  totalWorkUnits: number;
  completedWorkUnits: number;
  stage: "analyzing" | "writing" | "organizing";
}

export interface BaseOptions {
  dryRun: boolean;
  removeNamespaces: boolean;
  removeFauxNamespaces: boolean;
  organizeImports: boolean;
}

export interface ProcessProjectOpts extends BaseOptions {
  logger: Logger;
  updateState?: (data: Status) => void;
  additionalRenames?: Map<PackageName, PackageExportRename[]>;
}

/**
 * Process a project, removing namespaces and organizing imports.
 * @param project  The project to process
 * @param param1  Options
 * @returns
 */
export async function processProject(
  project: Project,
  {
    dryRun = false,
    logger,
    updateState = (_data) => undefined,
    additionalRenames,
    removeNamespaces,
    removeFauxNamespaces,
    organizeImports: shouldOrganizeImports,
  }: ProcessProjectOpts
) {
  dropDtsFiles(project);
  const context = new ProjectContext(project, logger);
  const totalFiles = project.getSourceFiles().length;
  // Three stages:
  // * analyzing:
  //   * if `removeNamespaces`: add `totalFiles`
  //   * if `additionalRenames`: add `totalFiles`
  // * organizing `[...changedFiles].length` work.
  // * writing  `[]
  function calculateTotalWorkUnits(changedFilesCount: number) {
    return (
      (removeNamespaces ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (dryRun ? 0 : changedFilesCount) +
      (shouldOrganizeImports ? changedFilesCount : 0)
    );
  }

  // Assume all files change for now
  let totalWorkUnits = calculateTotalWorkUnits(totalFiles);

  let completedWorkUnits = 0;
  // TODO: Rename totalFiles here to totalWorkUnits or similar
  updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });

  logger.debug("Running with opts %o", {
    dryRun,
    removeNamespaces,
    additionalRenames: [...(additionalRenames?.entries() ?? [])],
  });

  for (const sf of project.getSourceFiles()) {
    if (removeNamespaces) {
      calculateNamespaceRemovals(sf, context);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (removeFauxNamespaces) {
      const replacements = new ReplacementsWrapper(context);
      calculateNamespaceLikeRemovals(sf, replacements);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    if (additionalRenames) {
      const replacements: Replacement[] = [];
      addSingleFileReplacementsForRenames(sf, additionalRenames, replacements, logger);
      for (const r of replacements) {
        context.addReplacement(r);
      }
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }
  }

  // actually updates files in project!
  const changedFiles = [...processReplacements(project, context.getReplacements())];
  totalWorkUnits = calculateTotalWorkUnits(changedFiles.length);

  if (shouldOrganizeImports) {
    logger.debug("Organizing imports");
    updateState({ totalWorkUnits, completedWorkUnits, stage: "organizing" });
    organizeImportsOnFiles(project, changedFiles);
  }

  completedWorkUnits += changedFiles.length; // TODO make this granular?
  updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });

  if (dryRun) {
    logger.info("DRY RUN");
  } else {
    logger.info("Saving");
    await project.save();
    completedWorkUnits += changedFiles.length; // TODO make this granular?
    updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });
  }

  return {
    exportedRenames: context.getRecordedRenames(),
  };
}

export function getFilePath(filePath: string | Node) {
  return typeof filePath == "string" ? filePath : filePath.getSourceFile().getFilePath();
}
