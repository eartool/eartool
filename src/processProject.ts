import { Node, SyntaxKind, type Project, type SourceFile, NamedImports } from "ts-morph";
import { calculateNamespaceRemovals } from "./calculateNamespaceRemovals.js";
import type { Logger } from "pino";
import { ProjectContext } from "./Context.js";
import { processReplacements } from "./replacements/processReplacements.js";
import { dropDtsFiles } from "./utils/tsmorph/dropDtsFiles.js";
import { organizeImportsOnFiles } from "./utils/tsmorph/organizeImportsOnFiles.js";
import type { PackageExportRename } from "./replacements/PackageExportRename.js";
import type { PackageName } from "./PackageName.js";
import { addSingleFileReplacementsForRenames } from "./replacements/addSingleFileReplacementsForRenames.js";
import type { Replacement } from "./replacements/Replacement.js";
import * as Assert from "assert";
import { ReplacementsWrapper } from "./ReplacementsWrapper.js";
import type { Replacements } from "./replacements/Replacements.js";
import { isNamespaceLike } from "./utils/tsmorph/isNamespaceLike.js";
import { isAnyOf } from "@reduxjs/toolkit";

export interface Status {
  totalWorkUnits: number;
  completedWorkUnits: number;
  stage: "analyzing" | "writing" | "organizing";
}

export interface ProcessProjectOpts {
  dryRun?: boolean;
  logger: Logger;
  removeNamespaces: boolean;
  updateState?: (data: Status) => void;
  additionalRenames?: Map<PackageName, PackageExportRename[]>;
}

export async function processProject(
  project: Project,
  {
    dryRun = false,
    logger,
    updateState = (_data) => undefined,
    additionalRenames,
    removeNamespaces,
  }: ProcessProjectOpts
) {
  dropDtsFiles(project);
  const context = new ProjectContext(project, logger);
  const totalFiles = project.getSourceFiles().length;
  // Three stages:
  // * analyzing:
  //   * if `removeNamespaces`: add `totalFiles`
  //   * if `additionalRenames`: add `totalFiles`
  // * organizing `[...changedFiles].length` work.
  // * writing  `[]
  function calculateTotalWorkUnits(changedFilesCount: number) {
    return (
      (removeNamespaces ? totalFiles : 0) +
      (additionalRenames ? totalFiles : 0) +
      (dryRun ? 0 : changedFilesCount) +
      changedFilesCount
    );
  }

  // Assume all files change for now
  let totalWorkUnits = calculateTotalWorkUnits(totalFiles);

  let completedWorkUnits = 0;
  // TODO: Rename totalFiles here to totalWorkUnits or similar
  updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });

  logger.debug("Running with opts %o", {
    dryRun,
    removeNamespaces,
    additionalRenames: [...(additionalRenames?.entries() ?? [])],
  });

  for (const sf of project.getSourceFiles()) {
    if (removeNamespaces) {
      calculateNamespaceRemovals(sf, context);
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }

    const replacements = new ReplacementsWrapper(context);
    calculateNamespaceLikeRemovals(sf, replacements);

    if (additionalRenames) {
      const replacements: Replacement[] = [];
      addSingleFileReplacementsForRenames(sf, additionalRenames, replacements, logger);
      for (const r of replacements) {
        context.addReplacement(r);
      }
      completedWorkUnits++;
      updateState({ totalWorkUnits, completedWorkUnits, stage: "analyzing" });
    }
  }

  // actually updates files in project!
  const changedFiles = [...processReplacements(project, context.getReplacements())];
  totalWorkUnits = calculateTotalWorkUnits(changedFiles.length);

  logger.debug("Organizing imports");
  updateState({ totalWorkUnits, completedWorkUnits, stage: "organizing" });
  organizeImportsOnFiles(project, changedFiles);

  completedWorkUnits += changedFiles.length; // TODO make this granular?
  updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });

  if (dryRun) {
    logger.info("DRY RUN");
  } else {
    logger.info("Saving");
    await project.save();
    completedWorkUnits += changedFiles.length; // TODO make this granular?
    updateState({ totalWorkUnits, completedWorkUnits, stage: "writing" });
  }

  return {
    exportedRenames: context.getRecordedRenames(),
  };
}

export function getFilePath(filePath: string | Node) {
  return typeof filePath == "string" ? filePath : filePath.getSourceFile().getFilePath();
}

function calculateNamespaceLikeRemovals(sf: SourceFile, replacements: Replacements) {
  // TODO: Only perform task if its the only export
  // TODO: Should we check the filename too?
  // TODO: Check for collisions?

  for (const statement of sf.getStatements()) {
    if (!isNamespaceLike(statement)) continue;
    const varDecl = statement.getDeclarations()[0];

    const visitedSpecifiers = new Set();
    for (const refIdentifier of varDecl.findReferencesAsNodes()) {
      // alias import nodes show up twice for some reason
      // so we need to account for that

      const specifier = refIdentifier.getParentIf(
        isAnyOf(Node.isExportSpecifier, Node.isImportSpecifier)
      );
      if (!specifier) continue;
      if (visitedSpecifiers.has(specifier)) continue;

      const named = specifier.getParentIfOrThrow(isAnyOf(Node.isNamedExports, Node.isNamedImports));
      Assert.ok(named.getElements().length == 1);
      const varName = (specifier.getAliasNode() ?? specifier.getNameNode()).getText();
      replacements.replaceNode(named, `* as ${varName}`);
      visitedSpecifiers.add(specifier);
    }

    const syntaxList = varDecl.getInitializer().getExpression().getChildSyntaxList()!;

    // Drop `export const Name = {`
    replacements.remove(sf, statement.getStart(), syntaxList.getFullStart());

    // drop `} as const;`
    const closeBrace = syntaxList.getNextSiblingIfKindOrThrow(SyntaxKind.CloseBraceToken);
    replacements.remove(sf, closeBrace.getStart(), statement.getEnd());

    for (const q of varDecl.getInitializer().getExpression().getProperties()) {
      if (Node.isMethodDeclaration(q)) {
        replacements.insertBefore(q, "export function ");
      } else {
        replacements.addReplacement(
          sf,
          q.getStart(),
          q.getFirstChildByKindOrThrow(SyntaxKind.ColonToken).getEnd(),
          `export const ${q.getName()} = `
        );
      }
      replacements.removeNextSiblingIfComma(q);
    }
  }
}
