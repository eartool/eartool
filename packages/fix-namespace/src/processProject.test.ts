import { describe, expect, it } from "@jest/globals";
import { format } from "prettier";
import { Node, Project } from "ts-morph";
import { processProject, type ProcessProjectOpts } from "./processProject.js";
import { pino } from "pino";
import pinoPretty from "pino-pretty";
import * as Assert from "node:assert";

function formatTestTypescript(src: string) {
  return format(src, { parser: "typescript", tabWidth: 2, useTabs: false });
}

const cases: {
  name: string;
  inputs: Record<string, string>;
  additionalRenames?: ProcessProjectOpts["additionalRenames"];
}[] = [
  {
    name: "redeclare export",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export function bar() {
              return 5;
            }

            export function baz() {
              return 5;
            }

            export type Thing = string;

            export interface Props {
              what: Thing;
            }
          }
        `,

      "index.ts": `
          export {Foo} from "./foo";
      `,
    },
  },
  {
    name: "redeclare export w funky names",
    inputs: {
      "foo.tsx": `
          export namespace FooBar {
            export type ThingBar = string;

            export const doFoo = () => 5;

            export const VAR_BAR = 5;
          }
        `,

      "index.ts": `
          export {FooBar} from "./foo";
      `,
    },
  },
  {
    name: "redeclare export with ordering",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export interface Props {
              what: Thing;
            }
          }
        `,

      "index.ts": `
      export const before = 5;
          export {Foo} from "./foo";
          export const after = 5;
      `,
    },
  },
  {
    name: "doesnt clobber component exports",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
              export interface Props { what: number }
          }

          export class Foo extends React.Component<Foo.Props> {

          }
      `,
      "index.ts": `
          export {Foo} from "./foo";
      `,
    },
  },
  {
    name: "function invoke within namespace",
    inputs: {
      "foo.tsx": `
          export namespace Foo {
            export function bar() {
              baz();
            }

            export function baz() {
              return 5;
            }
          }
        `,
    },
  },
  {
    name: "combined types",
    inputs: {
      "foo.tsx": `
          export namespace AssociatedMapSection {
          export interface OgreProps {
            properties: Property<any>[];
          }
        
          export interface ReduxProps {
            appRealmId: RealmId;
            mapConfig: GaiaMapConfig;
          }
        
          export interface State {
            map?: MapSearchResult;
            user?: IAcmeUser;
            isLoading: boolean;
          }
        
          export type Props = OverviewObjectMinProps & OgreProps & ReduxProps;
      }`,
    },
  },
  {
    name: "works nicely with interfaces",
    // We cant break foo out in this case cause im too lazy to implement this another way.
    inputs: {
      "foo.tsx": `
          export namespace Foo {
              export interface Props { what: number }
          }

          class Foo extends React.Component<Foo.Props> {

          }
      `,
      "index.ts": `
          export {Foo} from "./foo";
      `,
    },
  },
  {
    name: "dont explode if error",
    // We cant break foo out in this case cause im too lazy to implement this another way.
    inputs: {
      "foo.ts": `
          const foo = 5;
          
          export namespace Wat {
              export const foo = 6;
          }
            `,
    },
  },
  {
    name: "simple",
    inputs: {
      "foo.ts": `
        const foo = 5;
        
        export namespace Wat {
          export const aasdf = 3;
          export const second = 5;

          export const thirdSpaced = 56;

          // Foo
          export const fourthWithComment = 555;
        }
        `,
    },
  },
  {
    name: "rename in other file",
    inputs: {
      "wat.ts": `
          export namespace Wat {
            export const key = 3;
            export function f() { return 5; }

            export class Foo {}

            export type Baz = string;
          }
        `,

      "refWat.ts": `
          import {Wat} from "./wat";

          console.log(Wat.key);
          console.log(Wat.f());
          console.log(new Wat.Foo());
          const f: Wat.Baz = "hi";
        `,
    },
  },
  {
    name: "Conflicting imports work",
    inputs: {
      "foo.ts": `
          export namespace Foo { export type Props = { hi: "mom" }}
      `,
      "bar.ts": `
          export namespace Bar { export type Props = { hi: "mom" }}
      `,
      "baz.ts": `
          import {Foo} from "./foo";
          import {Bar} from "./bar";
          export type Props = Foo.Props & Bar.Props;

      `,
    },
  },
  {
    name: "Simple additional renames work",
    inputs: {
      "foo.ts": `
          import {Foo} from "lib";
          export type MyProps = Foo.Props;
      `,
    },
    additionalRenames: new Map([["lib", [{ from: ["Foo", "Props"], to: ["FooProps"] }]]]),
  },
  {
    name: "Generics work",
    inputs: {
      "foo.ts": `
          export namespace MapElementViewerProperties {
            export interface OwnProps<T extends MapElement> {
              mapElement: T;
              locked: boolean;
              section: RightMapPanelType;
            }
          
            export interface StoreProps {
              mapDataState: MapDataState;
            }
          
            export interface NectarProps {
              dispatch: Dispatch;
            }
          
            export type Props<T extends MapElement> = OwnProps<T> & StoreProps & NectarProps;
          }
      `,
      "bar.ts": `
          import {MapElementViewerProperties} from "./foo";

          export type Foo<T extends MapElement> = MapElementViewerProperties.OwnProps<T>;
      `,
    },
  },
  {
    name: "Generics work with additional",
    inputs: {
      "foo.ts": `
          export namespace MapElementViewerProperties {
            export interface OwnProps<T extends MapElement> {
              mapElement: T;
              locked: boolean;
              section: RightMapPanelType;
            }
          
            export interface StoreProps {
              mapDataState: MapDataState;
            }
          
            export interface NectarProps {
              dispatch: Dispatch;
            }
          
            export type Props<T extends MapElement> = OwnProps<T> & StoreProps & NectarProps;
          }
      `,
      "bar.ts": `
          import {MapElementViewerProperties} from "./foo";
          import {Bleh} from "somelib"

          export type Foo<T extends MapElement & Bleh.Bar.Other> = MapElementViewerProperties.OwnProps<T>;
      `,
    },
    additionalRenames: new Map([
      ["somelib", [{ from: ["Bleh", "Bar", "Other"], to: ["Moo", "Cow"] }]],
    ]),
  },
  {
    name: "Handles const pattern nicely",
    inputs: {
      "index.ts": `
        export {Foo} from "./foo";
      `,
      "foo.ts": `
        export const Foo = {
          // direct function
          baz() {
            return 5; 
          },

          bar: () => {
            return 6;
          }
        } as const;
      `,
      "bar.ts": `
        import {Foo} from "./foo";

        console.log(Foo.bar());
        console.log(Foo.baz());
      `,
      "alias.ts": `
        import {Foo as Bar} from "./foo";

        console.log(Bar.bar());
        console.log(Bar.baz());
      `,
    },
  },
  {
    name: "Doesn't change as const that has constnats",
    inputs: {
      "index.ts": `
        export {Foo} from "./foo";
      `,
      "foo.ts": `
        export const Foo = {
          // direct function
          baz() {
            return 5; 
          },

          bar: () => {
            return 6;
          },

          FOO: 7
        } as const;
      `,
    },
  },
];

describe("processProject", () => {
  it.each(cases)("$name", async ({ inputs, additionalRenames }) => {
    const logger = createTestLogger();
    try {
      const project = createProjectForTest(inputs);

      await processProject(project, {
        logger,
        additionalRenames,
        removeFauxNamespaces: false,
        dryRun: false,
        organizeImports: true,
        removeNamespaces: true,
      });

      const fs = project.getFileSystem();

      const output = project
        .getSourceFiles()
        .map((sf) => {
          const filePath = sf.getFilePath();
          return formatTestTypescript(
            `
        //

        //
        // PATH: '${filePath}'
        //
        ${fs.readFileSync(filePath)}
        `
          );
        })
        .join();

      expect(output).toMatchSnapshot();
    } finally {
      logger.flush();
    }
  });

  it("records renames from root", async () => {
    const logger = createTestLogger();

    const project = createProjectForTest({
      "foo.ts": `
          export namespace Foo {
            export type Props = {};
          }
      `,
      "index.ts": `
          export {Foo} from "./foo";
      `,
    });

    const result = await processProject(project, {
      logger,
      removeNamespaces: true,
      removeFauxNamespaces: false,
      dryRun: false,
      organizeImports: true,
    });
    expect(result.exportedRenames).toHaveLength(1);
    expect(result.exportedRenames[0]).toEqual({ from: ["Foo", "Props"], to: ["FooProps"] });
  });
});

function createProjectForTest(inputs: Record<string, string>) {
  const project = new Project({
    useInMemoryFileSystem: true,
    skipAddingFilesFromTsConfig: true,
  });
  for (const [name, contents] of Object.entries(inputs)) {
    project.createSourceFile(name, formatTestTypescript(contents));
  }
  project.saveSync();
  return project;
}

function createTestLogger() {
  const { currentTestName } = expect.getState();
  Assert.ok(currentTestName != null);

  return pino(
    {
      level: "trace",
      serializers: {
        ...pino.stdSerializers,
        foo: (n: Node) => `${n.getSourceFile().getFilePath()}:${n.getStartLineNumber()}`,
      },
      hooks: {
        logMethod: function logMethod([msg, ...args], method, _foo) {
          args = args.map((a) => maybeConvertNodeToFileAndLineNum(a));
          // console.log([msg, ...args]);
          method.apply(this, [msg, ...args]);
        },
      },
    },
    pino.multistream([
      {
        level: "trace",
        stream: pino.destination({
          sync: true,
          mkdir: true,
          dest: `logs/processProject/${currentTestName}.log.json`,
        }),
      },
      {
        level: "trace",
        stream: pinoPretty.default({
          colorize: false,
          sync: true,
          singleLine: false,
          destination: pino.destination({
            sync: true,
            mkdir: true,
            append: false,
            dest: `logs/processProject/${currentTestName}.log.txt`,
          }),
        }),
      },
    ])
  );
}

function maybeConvertNodeToFileAndLineNum(a: any): any {
  if (a instanceof Node) {
    return `${a.getSourceFile().getFilePath()}:${a.getStartLineNumber()}`;
  }

  return a;
}
