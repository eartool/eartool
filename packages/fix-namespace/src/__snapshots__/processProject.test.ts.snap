// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processProject Conflicting imports work 1`] = `
"//

//
// PATH: '/bar.ts'
//

export type Props = { hi: "mom" };
,//

//
// PATH: '/baz.ts'
//
import { Props as BarProps } from "./bar";
import { Props as FooProps } from "./foo";
export type Props = FooProps & BarProps;
,//

//
// PATH: '/foo.ts'
//

export type Props = { hi: "mom" };
"
`;

exports[`processProject Doesn't change as const that has constnats 1`] = `
"//

//
// PATH: '/foo.ts'
//
export const Foo = {
  // direct function
  baz() {
    return 5;
  },

  bar: () => {
    return 6;
  },

  FOO: 7,
} as const;
,//

//
// PATH: '/index.ts'
//
export { Foo } from "./foo";
"
`;

exports[`processProject Generics work 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { OwnProps as MapElementViewerPropertiesOwnProps } from "./foo";

export type Foo<T extends MapElement> = MapElementViewerPropertiesOwnProps<T>;
,//

//
// PATH: '/foo.ts'
//

export interface OwnProps<T extends MapElement> {
  mapElement: T;
  locked: boolean;
  section: RightMapPanelType;
}

export interface StoreProps {
  mapDataState: MapDataState;
}

export interface NectarProps {
  dispatch: Dispatch;
}

export type Props<T extends MapElement> = OwnProps<T> &
  StoreProps &
  NectarProps;
"
`;

exports[`processProject Generics work with additional 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { Moo } from "somelib";
import { OwnProps as MapElementViewerPropertiesOwnProps } from "./foo";

export type Foo<T extends MapElement & Moo.Cow> =
  MapElementViewerPropertiesOwnProps<T>;
,//

//
// PATH: '/foo.ts'
//

export interface OwnProps<T extends MapElement> {
  mapElement: T;
  locked: boolean;
  section: RightMapPanelType;
}

export interface StoreProps {
  mapDataState: MapDataState;
}

export interface NectarProps {
  dispatch: Dispatch;
}

export type Props<T extends MapElement> = OwnProps<T> &
  StoreProps &
  NectarProps;
"
`;

exports[`processProject Handles const pattern nicely 1`] = `
"//

//
// PATH: '/alias.ts'
//
import * as Bar from "./foo";

console.log(Bar.bar());
console.log(Bar.baz());
,//

//
// PATH: '/bar.ts'
//
import * as Foo from "./foo";

console.log(Foo.bar());
console.log(Foo.baz());
,//

//
// PATH: '/foo.ts'
//

// direct function
export function baz() {
  return 5;
}

export const bar = () => {
  return 6;
};
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;

exports[`processProject Handles in file references to faux namespaces 1`] = `
"//

//
// PATH: '/foo.ts'
//

// direct function
export function baz() {
  return 5;
}

export const bar = () => {
  return baz();
};
,//

//
// PATH: '/index.ts'
//
export * as Foo from "./foo";
"
`;

exports[`processProject Multiple things in the file 1`] = `
"//

//
// PATH: '/bar.ts'
//
import { MetadataValue, foo } from "./foo";

console.log(MetadataValue.doThing());
console.log(foo());
,//

//
// PATH: '/foo.ts'
//
export namespace MetadataValue {
  export type Foo = string;
}

export const MetadataValue = {
  doThing() {
    return 5;
  },
};

export function foo() {
  return 5;
}
"
`;

exports[`processProject Properly handles conflict with module local variable or function 1`] = `
"//

//
// PATH: '/foo.ts'
//
function bar0() {
  return 5;
}
const baz0 = 5;

export function baz() {
  return baz0;
}

export const bar = () => {
  return bar0();
};
"
`;

exports[`processProject Renames import specifiers as needed 1`] = `
"//

//
// PATH: '/foo.ts'
//
import { bar as bar0 } from "lib";
import baz0 from "lib2";

// direct function
export function baz() {
  return bar0() + baz0();
}

export const bar = () => {
  return baz();
};
"
`;

exports[`processProject Renames star imports as needed 1`] = `
"//

//
// PATH: '/foo.ts'
//
import * as otherStar from "other";
import otherDefault, { other } from "other";

import * as bar0 from "lib";

// direct function
export function baz() {
  return bar0.baz() + other + otherStar.foo + otherDefault;
}

export const bar = () => {
  return baz();
};
"
`;

exports[`processProject Simple additional renames work 1`] = `
"//

//
// PATH: '/foo.ts'
//
import { FooProps } from "lib";
export type MyProps = FooProps;
"
`;

exports[`processProject combined types 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface OgreProps {
  properties: Property<any>[];
}

export interface ReduxProps {
  appRealmId: RealmId;
  mapConfig: GaiaMapConfig;
}

export interface State {
  map?: MapSearchResult;
  user?: IAcmeUser;
  isLoading: boolean;
}

export type Props = OverviewObjectMinProps & OgreProps & ReduxProps;
"
`;

exports[`processProject doesnt clobber component exports 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface Props {
  what: number;
}

export class Foo extends React.Component<Props> {}
,//

//
// PATH: '/index.ts'
//
export { Foo, type Props as FooProps } from "./foo";
"
`;

exports[`processProject dont explode if error 1`] = `
"//

//
// PATH: '/foo.ts'
//
const foo = 5;

export namespace Wat {
  export const foo = 6;
}
"
`;

exports[`processProject function invoke within namespace 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function barForFoo() {
  bazForFoo();
}

export function bazForFoo() {
  return 5;
}
"
`;

exports[`processProject redeclare export 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export function barForFoo() {
  return 5;
}

export function bazForFoo() {
  return 5;
}

export type ThingForFoo = string;

export interface Props {
  what: ThingForFoo;
}
,//

//
// PATH: '/index.ts'
//
export {
  barForFoo,
  bazForFoo,
  type Props as FooProps,
  type ThingForFoo,
} from "./foo";
"
`;

exports[`processProject redeclare export w funky names 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export type ThingFooBar = string;

export const doFooForFooBar = () => 5;

export const VAR_FOO_BAR = 5;
,//

//
// PATH: '/index.ts'
//
export { VAR_FOO_BAR, doFooForFooBar, type ThingFooBar } from "./foo";
"
`;

exports[`processProject redeclare export with ordering 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface Props {
  what: Thing;
}
,//

//
// PATH: '/index.ts'
//
export const before = 5;
export { type Props as FooProps } from "./foo";
export const after = 5;
"
`;

exports[`processProject rename in other file 1`] = `
"//

//
// PATH: '/refWat.ts'
//
import { BazForWat, FooForWat, fForWat, keyForWat } from "./wat";

console.log(keyForWat);
console.log(fForWat());
console.log(new FooForWat());
const f: BazForWat = "hi";
,//

//
// PATH: '/wat.ts'
//

export const keyForWat = 3;
export function fForWat() {
  return 5;
}

export class FooForWat {}

export type BazForWat = string;
"
`;

exports[`processProject simple 1`] = `
"//

//
// PATH: '/foo.ts'
//
const foo = 5;

export const aasdfForWat = 3;
export const secondForWat = 5;

export const thirdSpacedForWat = 56;

// Foo
export const fourthWithCommentForWat = 555;
"
`;

exports[`processProject works nicely with interfaces 1`] = `
"//

//
// PATH: '/foo.tsx'
//

export interface Props {
  what: number;
}

class Foo extends React.Component<Props> {}
,//

//
// PATH: '/index.ts'
//
export { type Props as FooProps } from "./foo";
"
`;
